### PR TITLE
add multiple routes and versions support

### PR DESCRIPTION
**Pull request by yinx**
This PR adds support for multiple swagger files on different routes or multiple versions on the same route.
Key features are:
- multiple routes can be defined in the config
- middleware config per route
- multiple possible versions

This PR contains breaking changes due to the overhaul in the config file.

### CODE REVIEW

File: src/SwaggerUiServiceProvider.php [hunk 1]
```diff
@@ -36,12 +37,13 @@ public function register() : void
...
     {
-        Route::middleware(config('swagger-ui.middleware', ['web', EnsureUserIsAuthorized::class]))
-            ->prefix(config('swagger-ui.path'))
-            ->group(function () {
-                Route::view('/', 'swagger-ui::index')->name('swagger-ui');
-
-                Route::get('openapi.json', OpenApiJsonController::class)->name('swagger-openapi-json');
-            });
+        collect(config('swagger-ui.files'))->each(function ($values) {
+            Route::middleware($values['middleware'] ?? ['web', EnsureUserIsAuthorized::class])
```
Comment (gdebrauwer) [inline src/SwaggerUiServiceProvider.php:41]:
```suggestion
            Route::middleware($values['middleware'])
```
I think we can remove the default middleware

File: config/swagger-ui.php [hunk 2]
```diff
@@ -3,73 +3,49 @@
...
+                'authorization_path' => 'oauth/authorize',
+
+                'client_id' => env('SWAGGER_UI_OAUTH_CLIENT_ID'),
+                'client_secret' => env('SWAGGER_UI_OAUTH_CLIENT_SECRET'),
+            ],
+
+            /*
+             * If enabled the file will be modified to set the server url and oauth urls.
+             */
+            'modify_file' => true,
```
Comment (gdebrauwer) [inline config/swagger-ui.php:48]: can we put this above the "oauth" config options?

File: src/SwaggerUiServiceProvider.php [hunk 3]
```diff
@@ -36,12 +37,13 @@ public function register() : void
...
+                ->group(function () use ($values) {
+                    Route::get(ltrim($values['path'], '/'), SwaggerViewController::class);
```
Comment (gdebrauwer) [inline src/SwaggerUiServiceProvider.php:43]:
```suggestion
                    Route::get($values['path'], SwaggerViewController::class);
```
trim is not needed here I think

File: src/Http/Controllers/OpenApiJsonController.php [hunk 4]
```diff
@@ -8,19 +8,26 @@
...
-        $json = $this->getJson();
+        $file = collect(config('swagger-ui.files'))->filter(function ($values) use ($filename, $path) {
+            return isset($values['versions'][$filename]) && ltrim($values['path'], '/') === $path;
+        })->first();
 
-        $json = $this->configureServer($json);
-        $json = $this->configureOAuth($json);
+        if (empty($file)) {
+            abort(404);
+        }
```
Comment (gdebrauwer) [inline src/Http/Controllers/OpenApiJsonController.php:19]: we can use firstOrFail-method on collection

File: src/Http/Controllers/SwaggerViewController.php [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+use Illuminate\Http\Request;
+
+class SwaggerViewController
+{
+    public function __invoke(Request $request)
+    {
+        $file = collect(config('swagger-ui.files'))->filter(function ($values) use ($request) {
+            return ltrim($values['path'], '/') === $request->path();
+        })->first();
```
Comment (gdebrauwer) [inline src/Http/Controllers/SwaggerViewController.php:13]:
```suggestion
        })->firstOrFail();
```
can you also add tests for this filtering logic

File: tests/OpenApiRouteTest.php [hunk 6]
```diff
@@ -38,11 +41,11 @@ public function openApiFileProvider() : array
      */
     public function it_sets_server_to_current_app_url_if_modify_file_is_enabled($openApiFile)
     {
-        config()->set('swagger-ui.file', $openApiFile);
-        config()->set('swagger-ui.modify_file', true);
+        config()->set('swagger-ui.files.0.versions', ['v1' => $openApiFile]);
```
Comment (gdebrauwer) [inline tests/OpenApiRouteTest.php:44]: when openApiFileProvider uses the "multiple versions" dataset, then this will create an invalid config. I don't think we should use the dataprovider for this

File: src/Http/Controllers/OpenApiJsonController.php [hunk 7]
```diff
@@ -8,19 +8,26 @@
 
 class OpenApiJsonController
 {
-    public function __invoke() : JsonResponse
+    public function __invoke(string $path, string $filename) : JsonResponse
     {
-        $json = $this->getJson();
+        $file = collect(config('swagger-ui.files'))->filter(function ($values) use ($filename, $path) {
```
Comment (gdebrauwer) [inline src/Http/Controllers/OpenApiJsonController.php:13]: can you add tests for this?

File: tests/SwaggerUiRouteTest.php [hunk 8]
```diff
@@ -26,23 +26,34 @@ protected function getPackageProviders($app) : array
...
+            ->assertSee('url: \'' . 'swagger2/v1' . '\'', false);
+    }
+
+    /** @test */
+    public function it_supports_multiple_verions()
+    {
+        $this->get('swagger-with-versions')
+            ->assertStatus(200)
+            ->assertSee('url: \'' . 'swagger-with-versions/v1' . '\'', false)
+            ->assertSee('url: \'' . 'swagger-with-versions/v2' . '\'', false);
```
Comment (gdebrauwer) [inline tests/SwaggerUiRouteTest.php:57]:
```suggestion
            ->assertSee('url: \'swagger-with-versions/v1\'', false)
            ->assertSee('url: \'swagger-with-versions/v2\'', false);
```

File: tests/TestCase.php [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+use Orchestra\Testbench\TestCase as BaseTestCase;
+
+abstract class TestCase extends BaseTestCase
+{
+    protected function defineEnvironment($app)
+    {
+        $app['config']->set('swagger-ui.files.1', config('swagger-ui.files.0'));
+        $app['config']->set('swagger-ui.files.1.oauth', ['client_id' => 1, 'client_secret' => 'foobar']);
+        $app['config']->set('swagger-ui.files.1.path', 'swagger2');
```
Comment (gdebrauwer) [inline tests/TestCase.php:13]: do we need this config? 🤔

**Review by gdebrauwer — CHANGES_REQUESTED** [review]

File: src/Http/Controllers/SwaggerViewController.php (hunk 5 above)
Comment (yinx) [reply src/Http/Controllers/SwaggerViewController.php:13]: Since the goal of this filter is to apply the correct config values to the route, isn't this already tested by the tests that check that the content of the page is modified according to the config values?

File: src/SwaggerUiServiceProvider.php [hunk 10]
```diff
@@ -36,12 +36,13 @@ public function register() : void
...
-                Route::view('/', 'swagger-ui::index')->name('swagger-ui');
-
-                Route::get('openapi.json', OpenApiJsonController::class)->name('swagger-openapi-json');
-            });
+        collect(config('swagger-ui.files'))->each(function ($values) {
+            Route::middleware($values['middleware'])
+                ->group(function () use ($values) {
+                    Route::get($values['path'], SwaggerViewController::class)->name($values['path'] . '.index');
+
+                    Route::get($values['path'] . '/{filename}', OpenApiJsonController::class)->name($values['path'] . '.json');
```
Comment (gdebrauwer) [inline src/SwaggerUiServiceProvider.php:44]: why did you change it to not using a path variable for first part of the route? manually now having to get the first segment of the route in the controller seems not ideal in my opinion

File: src/Http/Controllers/OpenApiJsonController.php (hunk 7 above)
Comment (gdebrauwer) [reply src/Http/Controllers/OpenApiJsonController.php:16]: can you add tests:
- it_returns_not_found_response_if_provided_version_does_not_exist_in_file
- it_returns_not_found_response_if_provided_file_does_not_exist (can you add this one for the swagger index route as well?)

**Review by gdebrauwer — CHANGES_REQUESTED** [review]

**Review by Bart-Kuenen-nextapps — APPROVED** [review]

File: src/Http/Controllers/OpenApiJsonController.php [hunk 11]
```diff
@@ -2,26 +2,42 @@
...
+    public function __invoke(Request $request, string $filename) : JsonResponse
     {
-        $json = $this->getJson();
+        $path = $request->segment(1);
+
+        try {
+            $file = collect(config('swagger-ui.files'))->filter(function ($values) use ($filename, $path) {
+                return isset($values['versions'][$filename]) && ltrim($values['path'], '/') === $path;
+            })->firstOrFail();
+        } catch (ItemNotFoundException $e) {
```
Comment (gdebrauwer) [inline src/Http/Controllers/OpenApiJsonController.php:22]:
```suggestion
        } catch (ItemNotFoundException) {
```

File: src/Http/Controllers/SwaggerViewController.php [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+class SwaggerViewController
+{
+    public function __invoke(Request $request)
+    {
+        try {
+            $file = collect(config('swagger-ui.files'))->filter(function ($values) use ($request) {
+                return ltrim($values['path'], '/') === $request->path();
+            })->firstOrFail();
+        } catch (ItemNotFoundException $e) {
+            return response()->json(['error' => 'File not found'], 404);
```
Comment (gdebrauwer) [inline src/Http/Controllers/SwaggerViewController.php:17]:
```suggestion
        } catch (ItemNotFoundException) {
            return abort(404);
```
seems a bit weird to return json response on the view route

File: tests/OpenApiRouteTest.php [hunk 13]
```diff
@@ -94,18 +92,37 @@ public function it_doesnt_sets_server_to_current_app_url_if_modify_file_is_disab
...
+
+    /** @test */
+    public function it_returns_not_found_response_if_provided_version_does_not_exist_in_file()
+    {
+        $this->getJson('swagger/v4')
+            ->assertStatus(404);
+    }
+
+    /** @test */
+    public function it_returns_not_found_response_if_provided_file_does_not_exist()
```
Comment (gdebrauwer) [inline tests/OpenApiRouteTest.php:116]:
```suggestion
    public function it_returns_not_found_response_if_provided_file_does_not_exist_even_when_provided_version_exists()
```

**Review by gdebrauwer — APPROVED** [review]
